Wikipedia/Code/WMFArticleViewController.m

```diff
@@ -97,7 +97,11 @@ - (nullable UIActivityViewController *)sharingActivityViewControllerWithTextSnip
     NSURL *url = [NSURL wmf_desktopURLForURL:self.url];
 
     if (url) {
-        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", url.absoluteString, @"wprov=sfsi1"]]];
+        NSString *urlAsString = url.absoluteString;
+        if([urlAsString hasSuffix:@"!"]) {
+            urlAsString = [urlAsString stringByReplacingOccurrencesOfString:@"!" withString:@"%21"];
+        }
+        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", urlAsString, @"wprov=sfsi1"]]];
```

Comment:
However, I think since `!` is actually allowed in the path and the bug is on twitter's end, we should merge this PR as is (since there are worthwhile fixes in it) and discuss whether or not we want to add the workaround for the twitter bug as a follow on (still tracked by https://phabricator.wikimedia.org/T100321).